Integrate the Smoke Tests Into Integration Test Framework To Allow Easier Extension

jonathanrainer opened the pull request:
This PR converts our existing smoke tests to not be run from `xtask` but instead using `rstest` as part of the integration testing framework. This makes it easier to extend in future and allows us to use all of rust's built in assertion libraries etc.

The one caveat with this that I would say is that the use of `#[once]` fixtures does mean that processes _will not_ get torn down properly in a non-CI context (i.e. running locally). This isn't brilliant but until `rstest` supports `afterTest` hooks or similar, which might be a while away: https://github.com/la10736/rstest/issues/197#issuecomment-1610897815 there's not a lot we can do.

If we're running them in CI only it's not such a huge problem, but we'll need to be aware of it as the test suite grows. 

Successful run: https://github.com/apollographql/rover/actions/runs/10073378731/job/27847284362

jonathanrainer committed:
ROVER-98 Move from xtask to integration tests

aaronArinder commented on tests/e2e/dev.rs:
```diff
@@ -0,0 +1,92 @@
+#[case::multiple_subgraphs("query {order(id: \"order:2\") { items { product { id } inventory { inventory } colorway } buyer { id } } }", json!({"data":{"order":{"items":[{"product":{"id":"product:1"},"inventory":{"inventory":0},"colorway":"Red"}],"buyer":{"id":"user:1"}}}}))]
+#[case::deprecated_field("query {product(id: \"product:2\") { reviews { author id } } }", json!({"data":{"product":{"reviews":[{"author":"User 1","id":"review:2"},{"author":"User 1","id":"review:7"}]}}}))]
+#[case::deprecated_introspection("query {__type(name:\"Review\"){ fields(includeDeprecated: true) { name isDeprecated deprecationReason } } }", json!({"data":{"__type":{"fields":[{"name":"id","isDeprecated":false,"deprecationReason":null},{"name":"body","isDeprecated":false,"deprecationReason":null},{"name":"author","isDeprecated":true,"deprecationReason":"Use the new `user` field"},{"name":"user ...
+#[ignore]
```
are you wanting this `#[ignore]` here?

jonathanrainer replied:
Yes, I want these tests to only be invoked when explicitly asked for, which in this case means passing the `--ignored` flag to the binary. Otherwise we'll be tying up the unit tests in CircleCI with a lot of stuff that it's not configured for.

jonathanrainer committed:
ROVER-98 Run smoke tests in the CI/CD pipeline

jonathanrainer committed:
ROVER-98 Try a less complicated approach

Saving the binary name upfront is a bit silly.
It's easier to just save the artifacts as we find
them and then look for them when we invoke.

jonathanrainer committed:
ROVER-98 Re-add Volta and make `rover dev` #[once]

Running multiple instances of `rover dev` is silly when we're
testing its interactions. In future, we could run multiple instances
if we want to, but it's just a cause of flakes for now.

jonathanrainer committed:
ROVER-98 Don't rely on commands in the repo

They won't work within Powershell so implement the command
ourselves.

jonathanrainer commented on tests/e2e/dev.rs:
```diff
@@ -0,0 +1,98 @@
+            .await
+            .expect("foo")
+        })
+    })
```
I'll admit I stole this idiom from a similar solution to the problem of `rstest` not supporting `async` `#[once]` fixtures. There might be a better way to do this and if so then I'm all ears!

jonathanrainer commented on tests/e2e/mod.rs:
```diff
@@ -0,0 +1,118 @@
+            })
+        })
+        .expect("Could not execute connectivity check");
+    }
```
I appreciate this isn't the most efficient if you have a very large number of subgraphs, I tried mapping the `subgraph_urls` into a Vec of Future but couldn't make it all hang together. Again if there's a better way I'm all ears!

loshz replied:
Is it worth us picking a random subset instead? How many URLs do we envisage here? Is it likely to grow?

nmoutschen replied:
This is only once for tests, so I'm not too concerned about the performance just yet 👀

jonathanrainer replied:
Also it's pulling from a specific example at which the rate of change is pretty glacial, so I'm not enormously worried

jonathanrainer commented on tests/e2e/mod.rs:
```diff
@@ -0,0 +1,118 @@
+                    url, e
+                );
+            }
+        }
```
My brain keeps prodding me that this could be better but I can't for the life of me think how! 🤦

loshz replied:
Do we want to add a backoff or at least a timeout?

jonathanrainer replied:
It gets timed out because I call it inside of a tokio `timeout` closure, so that's enforced at a higher level. But I'm wondering if it's better to enforce that lower down so you don't have to remember to do it when this is called... I'll do that

nmoutschen replied:
+1 on timeout here so we don't have an infinite loop

loshz commented on Cargo.toml:
```diff
@@ -127,7 +131,7 @@ thiserror = "1"
 termimad = "0.29"
 tempfile = "3.8"
-tokio = "1.32"
+tokio = "1.38.1"
```
Do we want to enforce the path version? (Unsure if cargo holds us to this?)

loshz commented on tests/e2e/mod.rs:
```diff
@@ -0,0 +1,118 @@
+                }
+            }
+            Err(e) => {
+                println!(
```
```suggestion
                eprintln!(
```

nmoutschen commented on tests/e2e/dev.rs:
```diff
@@ -0,0 +1,98 @@
+const ROUTER_PORT: u32 = 4123;
+
+#[fixture]
+#[once]
+fn run_rover_dev(#[from(run_subgraphs_retail_supergraph)] working_dir: &TempDir) -> String {
```
Should we have a bespoke return type that implements `Drop` here?

jonathanrainer replied:
Discussed in Slack here: https://apollograph.slack.com/archives/C06HGDC3W3Z/p1721901865421839

jonathanrainer committed:
ROVER-98 Respond to PR comments
